### PR TITLE
Add imports / exports lint rules

### PR DESCRIPTION
**Pull request by SuperSodaSea**
It's time to sort the imports and exports!

#### Description of change

ESLint plugin `eslint-plugin-import` and `eslint-plugin-simple-import-sort` are added.

The following ESLint rules are added or modified:

```js
{
    "no-multi-spaces": [1, { "exceptions": {} }],
    "object-curly-newline": [1, { "multiline": true, "consistent": true }],

    "import/first": 1,
    "import/newline-after-import": 1,
    "simple-import-sort/imports": [1, {
        "groups": [
            [
                "^\\w", // External
                "^\\u0000", // External side effect
                "^@pixi/", // @pixi/*
                "^\\u0000@pixi/", // @pixi/* side effect
                "^\\.", // Relative
                "^\\u0000\\.", // Relative side effect
                "^" // Others
            ],
            [
                "^\\w.*\\u0000$", // External types
                "^@pixi/.*\\u0000$", // @pixi/* types
                "^\\..*\\u0000$", // Relative types
                "\\u0000$" // Other types
            ]
        ]
    }],
    "simple-import-sort/exports": 1
}
```

You can see the effect through [Files changed](https://github.com/pixijs/pixijs/pull/8932/files). Most of them are done by `npm run lintfix`, with a few manual fixes.

If you have any idea about the rules, or something is broken after sorting imports, comments are welcomed!

Note: The change in `.eslintrc.json` is temporary, when everything is ready, the rules will be added to [`@pixi/eslint-config`](https://github.com/pixijs/eslint-config).

#### Pre-Merge Checklist

- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

### CODE REVIEW

File: bundles/pixi.js-node/src/index.ts [hunk 1]
```diff
@@ -10,10 +10,7 @@ import { NoiseFilter } from '@pixi/filter-noise';
 import '@pixi/mixin-cache-as-bitmap';
 import '@pixi/mixin-get-child-by-name';
 import '@pixi/mixin-get-global-position';
-// eslint-disable-next-line @typescript-eslint/no-duplicate-imports
 import { NodeCanvasResource } from './adapter';
```
Comment (SuperSodaSea) [inline bundles/pixi.js-node/src/index.ts:13]: The `import './adapter';` and `import { NodeCanvasResource } from './adapter';` is manually changed. It is not necessary to import adapter ahead of everything in my opinion.

File: packages/assets/src/detections/parsers/index.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
 export * from './detectAvif';
 export * from './detectWebp';
+// Defaults should be at the last
 export * from './detectDefaults';
```
Comment (SuperSodaSea) [inline packages/assets/src/detections/parsers/index.ts:4]: If the order matters, we can add a comment like this to avoid sorting.

**Review by bigtimebuddy — APPROVED** [review]
I love this change. I'm glad we are getting more tidy about imports. And I'm okay moving the import/export rules to `@pixi/eslint-config`.

**Comment by bigtimebuddy** [comment]
@Zyie would you might weighing-in on this change?

File: packages/core/test/UniformBuffer.tests.ts [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
-    generateProgram,
-    getTestContext
-} from '@pixi/core';
+    Shader,
+    UniformGroup } from '@pixi/core';
+import { Matrix, Point, Rectangle } from '@pixi/math';
 
-import { Matrix, Rectangle, Point } from '@pixi/math';
+import type {
+    IRenderingContext } from '@pixi/core';
```
Comment (Zyie) [inline packages/core/test/UniformBuffer.tests.ts:13]: While we are looking at linting imports/exports, this is something that has always bugged me

Is there a way to lint this sort of thing where clearly the import would fit on 1 line?

File: packages/core/src/shader/Program.ts [hunk 4]
```diff
@@ -1,9 +1,8 @@
-import { setPrecision,
-    getMaxFragmentPrecision } from './utils';
-import { ProgramCache, isMobile } from '@pixi/utils';
+import { PRECISION } from '@pixi/constants';
+import { isMobile, ProgramCache } from '@pixi/utils';
 import defaultFragment from './defaultProgram.frag';
 import defaultVertex from './defaultProgram.vert';
-import { PRECISION } from '@pixi/constants';
+import {     getMaxFragmentPrecision, setPrecision } from './utils';
```
[the remark this answers is not in the document]
Comment (SuperSodaSea) [reply packages/core/src/shader/Program.ts:5]: https://github.com/pixijs/eslint-config/blob/aa269b8dda4a5bceec47f1fa44632f17bac9851b/index.js#L90

```json
"no-multi-spaces": [1, { "exceptions": { "VariableDeclarator": true, "ImportDeclaration": true, "Property": true } }],
```

Seems that `no-multi-spaces` is disabled for `import` for some reason. Maybe we can change this.

File: packages/core/test/UniformBuffer.tests.ts (hunk 3 above)
Comment (SuperSodaSea) [reply packages/core/test/UniformBuffer.tests.ts:13]: https://github.com/pixijs/eslint-config/blob/main/index.js#L202

```json
"object-curly-newline": 0,
```

This one is related to [`object-curly-newline`](https://eslint.org/docs/latest/rules/object-curly-newline#importdeclaration-and-exportdeclaration). Maybe we can add this rule.

**Comment by SuperSodaSea** [comment]
@Zyie These rules are applied in 862449d1e0cc613510d472249d908fbb272f582e to solve the problem you have mentioned:

```json
"no-multi-spaces": [1, { "exceptions": {} }],
"object-curly-newline": [1, { "multiline": true, "consistent": true }],
```

---

@bigtimebuddy @Zyie By the way, what's your opinion about changing [`comma-dangle`](https://eslint.org/docs/latest/rules/comma-dangle) from `only-multiline` to `always-multiline`? This can force a comma for multiline `{}` / `[]` / `()`, making the code style more consistant.

```diff
      settings,
-     utils
+     utils,
  } from '@pixi/core';
```

**Comment by bigtimebuddy** [comment]
I don't like comma dangle on arrays and function args, I find this confusing for undefined last item.

**Comment by Julien-Marcou** [comment]
> I don't like comma dangle on arrays and function args, I find this confusing for undefined last item.

For imports I don't really care as they are mostly managed automatically by the IDE or the linter, but for objects and arrays, I do like the dangling commas, they make the commits cleaner and reduce the chances of conflicts.

**Comment by SuperSodaSea** [comment]
> I don't like comma dangle on arrays and function args, I find this confusing for undefined last item.

They can be set seperately, for example:

```json
"comma-dangle": [1, {
    "arrays": "always-multiline",
    "objects": "always-multiline",
    "imports": "always-multiline",
    "exports": "always-multiline",
    "functions": "never"
}],
```

**Comment by SuperSodaSea** [comment]
If anyone has different ideas about the `comma-dangle` rule, feel free to comment! Or we can just leave it to another PR.

**Comment by bigtimebuddy** [comment]
I'd like to get this one in. Let's do comma-dangle as another PR.

**Comment by SuperSodaSea** [comment]
> There's a few more things I'd like to move to eslint-config that are currently in this eslintrc (some other typescript rules, the comma-dangle stuff). So my preference is to do the eslint-config release and bump once after everything is settled.

@bigtimebuddy That make sense! :smile:

**Pull request merged**